lib/Backend/LowerMDSharedSimd128.cpp

```diff
@@ -1042,88 +1155,178 @@ IR::Instr* LowererMD::Simd128LowerStoreElem(IR::Instr *instr)
+        break;
+    default:;
+        Assume(UNREACHED);
     }
     instr->Remove();
     return instrPrev;
 }
 
+void
+LowererMD::Simd128GenerateUpperBoundCheck(IR::RegOpnd *indexOpnd, IR::IndirOpnd *indirOpnd, ValueType arrType, IR::Instr *instr)
+{
+    Assert(!m_func->m_workItem->GetFunctionBody()->GetIsAsmjsMode());
+
+    IR::ArrayRegOpnd *arrayRegOpnd = indirOpnd->GetBaseOpnd()->AsArrayRegOpnd();
+    IR::Opnd* headSegmentLengthOpnd;
+
+    if (arrayRegOpnd->EliminatedUpperBoundCheck())
+    {
+        // already eliminated or extracted by globOpt (OptArraySrc). Nothing to do. 
+        return;
+    }
+
+    if (arrayRegOpnd->HeadSegmentLengthSym())
+    {
+        headSegmentLengthOpnd = IR::RegOpnd::New(arrayRegOpnd->HeadSegmentLengthSym(), TyUint32, m_func);
+    }
+    else
+    {
+        // (headSegmentLength = [base + offset(length)])
+        int lengthOffset;
+        lengthOffset = m_lowerer->GetArrayOffsetOfLength(arrType);
+        headSegmentLengthOpnd = IR::IndirOpnd::New(arrayRegOpnd, lengthOffset, TyUint32, m_func);
+    }
+
+    IR::LabelInstr * skipLabel = Lowerer::InsertLabel(false, instr);
+    int32 elemCount = Lowerer::SimdGetElementCountFromBytes(arrayRegOpnd->GetValueType(), instr->dataWidth);
+    if (indexOpnd)
+    {
+        //  MOV tmp, elemCount
+        //  ADD tmp, index
+        //  CMP tmp, Length  -- upper bound check
+        //  JBE  $storeLabel
+        //  Throw RuntimeError
+        //  skipLabel:
+        IR::RegOpnd *tmp = IR::RegOpnd::New(indexOpnd->GetType(), m_func);
+        IR::IntConstOpnd *elemCountOpnd = IR::IntConstOpnd::New(elemCount, TyInt8, m_func, true);
+        m_lowerer->InsertMove(tmp, elemCountOpnd, skipLabel);
+        Lowerer::InsertAdd(false, tmp, tmp, indexOpnd, skipLabel);
+        m_lowerer->InsertCompareBranch(tmp, headSegmentLengthOpnd, Js::OpCode::BrLe_A, true, skipLabel, skipLabel);
+    }
+    else
+    {
+        // CMP Length, (offset + elemCount)
+        // JA $storeLabel
+        int32 offset = indirOpnd->GetOffset();
+        int32 index = offset + elemCount;
+        m_lowerer->InsertCompareBranch(headSegmentLengthOpnd, IR::IntConstOpnd::New(index, TyInt32, m_func, true), Js::OpCode::BrLe_A, true, skipLabel, skipLabel);
+    }
+    m_lowerer->GenerateRuntimeError(skipLabel, JSERR_ArgumentOutOfRange, IR::HelperOp_RuntimeRangeError);
+    return;
+}
+
+void
+LowererMD::Simd128LoadHeadSegment(IR::IndirOpnd *indirOpnd, ValueType arrType, IR::Instr *instr)
+{
+
+    // For non-asm.js we check if headSeg symbol exists, else load it.
+    IR::ArrayRegOpnd *arrayRegOpnd = indirOpnd->GetBaseOpnd()->AsArrayRegOpnd();
+    IR::RegOpnd *headSegmentOpnd;
+    
+    if (arrayRegOpnd->HeadSegmentSym())
+    {
+        headSegmentOpnd = IR::RegOpnd::New(arrayRegOpnd->HeadSegmentSym(), TyMachPtr, m_func);
+    }
+    else
+    {
+        // REVIEW: Is this needed ? Shouldn't globOpt make sure headSegSym is set and alive ?
```

Comment:
An ArrayRegOpnd for a LdElem or StElam should always have a headSegSym.  It just doesn't create it for a LdLen.